### PR TITLE
98 add remove variant

### PR DESCRIPTION
**Pull request by andrii-kovalenko-ct**
#### Summary
In `ProductVariantUpdateActionUtils` added:
  - `buildRemoveVariantUpdateAction()`
  - `buildAddVariantUpdateAction()`
  - `buildChangeMasterVariantUpdateAction()`

Respective unit tests added with Payment mock files.

#### Relevant Issues
  
#98

#### Todo

- [ ] more unit tests for corner cases?

#### Hints for Review

`ProductVariantUpdateActionUtils` is the main implementation

`CollectionUtils` is additional utilities class for filtering and converting collections

### CODE REVIEW

File: src/main/java/com/commercetools/sync/products/utils/ProductVariantUpdateActionUtils.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+import static com.commercetools.sync.commons.utils.CollectionUtils.filterCollection;
+import static com.commercetools.sync.commons.utils.CommonTypeUpdateActionUtils.buildUpdateAction;
+import static java.util.stream.Collectors.toList;
+
+public final class ProductVariantUpdateActionUtils {
+
+    /**
+     * <b>Note:</b> if you do both add/remove product variants - <b>first always remove the variants</b>,
+     * and then - add.
+     * The reason of such restriction: if you add first you could have duplication exception on the keys,
```
Comment (heshamMassoud) [inline src/main/java/com/commercetools/sync/products/utils/ProductVariantUpdateActionUtils.java]: > if you add first, the update action could fail due to having a duplicate `key` or `sku` with variants which were expected to be removed anyways. So issuing remove update action first will fix such issue.

File: src/main/java/com/commercetools/sync/products/utils/ProductVariantUpdateActionUtils.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+     *     <li>attributes</li>
+     *     <li>prices</li>
+     *     <li>images</li>
+     * </ul>
+     *
+     * @param draft {@link ProductVariantDraft} which to add.
+     * @return new {@link AddVariant} update action with properties from {@code draft}
+     */
+    @Nonnull
+    public static AddVariant buildAddVariantUpdateActionFromDraft(@Nonnull final ProductVariantDraft draft) {
```
Comment (heshamMassoud) [inline src/main/java/com/commercetools/sync/products/utils/ProductVariantUpdateActionUtils.java]: would be good to add the JVM SDK :)

**Comment by heshamMassoud** [comment]
@andrii-kovalenko-ct Also this PR is missing using the actual implemented update action methods in the ProductSyncUtils#buildCoreActions. So that the sync process can actually use them. Please add them there and add at least one IT that tests syncing a product with different variants and check if the product was updated and synced correctly.

File: src/main/java/com/commercetools/sync/products/utils/ProductVariantUpdateActionUtils.java (hunk 2 above)
Comment (heshamMassoud) [reply src/main/java/com/commercetools/sync/products/utils/ProductVariantUpdateActionUtils.java]: also why `public`?

File: src/main/java/com/commercetools/sync/products/utils/ProductVariantUpdateActionUtils.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+     * @param newProduct new product draft with variants <b>with resolved references prices references</b>
+     * @return list of update actions to add new variants.
+     * @see #buildRemoveVariantUpdateAction(Product, ProductDraft)
+     */
+    @Nonnull
+    public static List<UpdateAction<Product>> buildAddVariantUpdateAction(@Nonnull final Product oldProduct,
+                                                                          @Nonnull final ProductDraft newProduct) {
+        final List<ProductVariantDraft> newVariants = newProduct.getVariants();
+        final Set<String> oldVariantKeys =
+                collectionToSet(oldProduct.getMasterData().getStaged().getVariants(), ProductVariant::getKey);
```
Comment (heshamMassoud) [inline src/main/java/com/commercetools/sync/products/utils/ProductVariantUpdateActionUtils.java]: @andrii-kovalenko-ct `getVariants()` doesn't include the master variant. `getAllVariants()` does. So It means that the master variant will never be synced this way! 😮

**Comment by andrii-kovalenko-ct** [comment]
>Please move the JSON resources to the resources directory. Other than that looks good :)

but they are in the resources directory. Which JSON you mean?

File: src/main/java/com/commercetools/sync/products/utils/ProductVariantUpdateActionUtils.java (hunk 1 above)
Comment (andrii-kovalenko-ct) [reply src/main/java/com/commercetools/sync/products/utils/ProductVariantUpdateActionUtils.java]: 03193bf

File: src/main/java/com/commercetools/sync/products/utils/ProductVariantUpdateActionUtils.java (hunk 2 above)
Comment (andrii-kovalenko-ct) [reply src/main/java/com/commercetools/sync/products/utils/ProductVariantUpdateActionUtils.java]: c05f147
made internal to use in the tests.

**Comment by heshamMassoud** [comment]
> but they are in the resources directory. Which JSON you mean?

Ah ok, I didn't notice

File: src/main/java/com/commercetools/sync/commons/utils/CommonTypeUpdateActionUtils.java [hunk 4]
```diff
@@ -19,27 +19,46 @@
...
+     * @param <S>                  the type of the objects to compare
+     * @param <U>                  certain {@link UpdateAction} implementation type
+     * @return A filled optional with the update action or an empty optional if the object values are identical
      */
     @Nonnull
-    public static <T, S> Optional<UpdateAction<T>> buildUpdateAction(@Nullable final S oldObject,
-                                                                  @Nullable final S newObject,
-                                                                  @Nonnull final Supplier<UpdateAction<T>>
-                                                                    updateActionSupplier) {
+    public static <T, S, U extends UpdateAction<T>> Optional<U> buildUpdateAction(
```
Comment (heshamMassoud) [inline src/main/java/com/commercetools/sync/commons/utils/CommonTypeUpdateActionUtils.java]: @andrii-kovalenko-ct Why is this better?
